### DISCUSSION TITLE
Invoke long running Google Cloud function from DAG

### DISCUSSION THREAD

**Discussion by iblessedi** [apache/airflow#23264] [Q&A]
Hello! I am using CloudFunctionInvokeFunctionOperator and I am invoking a google cloud function. 
```
    CloudFunctionInvokeFunctionOperator(
        task_id='invoke_function',
        dag=reprocess_dag,
        function_id=config['function_name'],
        location=config['function_location'],
        gcp_conn_id="google_cloud_default",
        project_id=config['project_id'],
        input_data={}
    )
```

If I trigger a function which executes fast ( < 1 minute) - everything works. However if the function runs for longer than a minute - airflow thinks that's an error and makes a retry.
```
[2022-04-26 15:44:33,892] {http.py:171} WARNING - Sleeping 1.90 seconds before retry 1 of 5 for request: POST https://cloudfunctions.googleapis.com/v1/projects/project:call?alt=json, after The read operation timed out
[2022-04-26 15:45:36,200] {http.py:171} WARNING - Sleeping 2.02 seconds before retry 2 of 5 for request: POST https://cloudfunctions.googleapis.com/v1/projects/project:call?alt=json, after The read operation timed out
[2022-04-26 15:46:38,624] {http.py:171} WARNING - Sleeping 1.34 seconds before retry 3 of 5 for request: POST https://cloudfunctions.googleapis.com/v1/projects/project:call?alt=json, after The read operation timed out
[2022-04-26 15:47:40,325] {http.py:171} WARNING - Sleeping 13.79 seconds before retry 4 of 5 for request: POST https://cloudfunctions.googleapis.com/v1/projects/project:call?alt=json, after The read operation timed out

```

How can I tell Airflow that it need to wait a little longer for the response from the GCP? I tried timeout, execution_timeout and no luck. Thanks!

**Comment by potiuk** [comment]
1 minute is default for Cloud Function https://cloud.google.com/functions/docs/concepts/exec  - I think there is no feature to extend it in the hook but feel free to add it as contribution.

**Reply by rajeshdhanda** [reply]
Hi @potiuk, I didn't find any 1 minute limit in the link you have provided. I have manually deployed a cloud function with 540 seconds as the timeout, and invoking the cloud function as 
```
    invoke_task = CloudFunctionInvokeFunctionOperator(
        task_id="invoke_task",
        project_id=GCP_PROJECT_ID,
        location=GCP_LOCATION,
        execution_timeout=timedelta(seconds=540),
        input_data={"data": json.dumps()},   
        function_id=GCF_SHORT_FUNCTION_NAME,
        gcp_conn_id="cf_conn" )
```
but still ```execution_timeout``` is not working and the task failed in 1 minute. 
Although  ```execution_timeout```  works fine in other operators like ```BashOperator```.

**Reply by potiuk** [reply]
Maybe it's changed in the meantime ? Things could have changed in the meanitme? Maybe you have some old configuration or account? or mayb your cloud function has old limits set. No idea. I think you should raise question to Google Cloud support if you need asnwer.

**Comment by iblessedi** [comment]
Yes, that's the default value, but that's ok that function may run longer...

**Comment by joao-maia-maersk** [comment]
Hello guys, I'm facing the same problem. Any solution here?

**Reply by potiuk** [reply]
I do not think so - there is a 10 minute limit in Google (see the link I posted). If you hit it, then you should see if you can extend it. If your post gets cancelled before then you need to check what's cancelling it - what are the logs of yours @joao-maia-maersk ?

**Reply by joao-maia-maersk** [reply]
Hello @potiuk, I managed to find a solution to this problem. I changed a parameter in the http.py file from the googleapiclient package. 
The parameter is the DEFAULT_HTTP_TIMEOUT_SEC. So you can change it in order to fulfill your needs.

**Reply by potiuk** [reply]
Good to know! But rather you should direct it to @iblessedi @joao-maia-maersk  -> if anyone would  like to add it as a PR to be able to change it Airflow - feel free :)

**Reply by megha-compass** [reply]
@joao-maia-maersk Thanks for the solution. Could you please elaborate where exactly you modified/mentioned the updated DEFAULT_HTTP_TIMEOUT_SEC parameter? Is it somewhere in Airflow DAG? If yes, where exactly?

**Comment by ellie-xr-zhu** [comment]
Hello, I have the same issue. Is there any fix now? thank you!

**Comment by rajeshdhanda** [comment]
Using SimpleHttpOperator

```
import google.oauth2.id_token
import google.auth.transport.requests

request = google.auth.transport.requests.Request()
audience = f'https://{GCP_LOCATION}-{GCP_PROJECT_ID}.cloudfunctions.net/{CLOUD_FUNCTION_NAME}'
TOKEN = google.oauth2.id_token.fetch_id_token(request, audience)

invoke_task = SimpleHttpOperator(
        task_id= 'invoke_cf',
        method='POST',
        http_conn_id='http',
        endpoint=CLOUD_FUNCTION_NAME,
        execution_timeout=timedelta(seconds=600),
        headers={'Authorization': f"Bearer {TOKEN}", "Content-Type": "application/json"},
        data=json.dumps({}), 
        )

```
